File: src/20.有效的括号.cs
/*
 * @lc app=leetcode.cn id=20 lang=csharp
 *
 * [20] 有效的括号
 */

// @lc code=start
using System.Collections;
using System.Collections.Generic;
public class Solution
{
    public bool IsValid(string s)
    {
        var dic = new Dictionary<char, char>()
            {
                {'(',')'},
                {'[',']'},
                {'{','}'}
            };
        var stack = new Stack<char>();
        for (int i = 0; i < s.Length; i++)
        {
            if (!stack.Any())
            {
                stack.Push(s[i]);
                continue;
            }
            if (dic.ContainsKey(stack.Peek()))
            {
                if (s[i] != dic[stack.Peek()]) stack.Push(s[i]);
                else stack.Pop();
            }
        }
        if (stack.Any())
        {
            return false;
        }
        return true;
    }
}
// @lc code=end

File: src/20.有效的括号.java
import java.util.Stack;

/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char item = s.charAt(i);
            if (st.isEmpty()){
                st.push(item);
                continue;
            }
            if (item == '(') {
                st.push(item);
            } else if (item == ')') {
                if (st.peek() == '(') {
                    st.pop();
                    continue;
                }
                return false;
            } else if (item == '[') {
                st.push(item);
            } else if (item == ']') {
                if (st.peek() == '[') {
                    st.pop();
                    continue;
                }
                return false;
            } else if (item == '{') {
                st.push(item);
            } else if (item == '}') {
                if (st.peek() == '{') {
                    st.pop();
                    continue;
                }
                return false;
            }
        }
        return st.isEmpty();
    }
}
// @lc code=end
